README.md
# timm [![Build Status](https://github.com/guigrpa/timm/workflows/Tests/badge.svg)](https://github.com/guigrpa/timm/actions) [![Coverage Status](https://coveralls.io/repos/github/guigrpa/timm/badge.svg?branch=master)](https://coveralls.io/github/guigrpa/timm?branch=master) [![npm version](https://img.shields.io/npm/v/timm.svg)](https://www.npmjs.com/package/timm)

Immutability helpers with fast reads and acceptable writes ([blog post](http://guigrpa.github.io/2016/06/16/painless-immutability/))


## Installation

```
$ npm install --save timm
```


## Motivation

I know, I know... the world doesn't need yet another immutability library, especially with the likes of [ImmutableJS](http://facebook.github.io/immutable-js/) and [seamless-immutable](https://github.com/rtfeldman/seamless-immutable) around.

And yet... I felt the urge, at least just to cover my limited needs. ImmutableJS is a solid, comprehensive and highly-performant solution, but this power comes at a price: mixing up ImmutableJS's Maps and Lists with your plain objects can cause some friction, and reading those objects (in my case, more often than writing them) isn't that convenient.

On the other side, *seamless-immutable* solves the "friction" problem by using plain objects and arrays, but seems to have some performance issues (at least in my benchmarks, see below).

*timm*'s approach: use plain objects and arrays and provide simple mutation functions to handle most common operations (suggestions are welcome!). As a bonus, *timm* creates new objects *lazily*, when it confirms that the operation will mutate the input object; in other words, **operations that don't modify an object always return the object itself**. This alleviates pressure on the garbage collector and lets you easily check whether an object changed after an operation: `merge(obj, { a: 3 }) === obj`.

**Important notice:** *timm* does *not* freeze the objects it provides. In other words, it doesn't protect you against inadvertently modifying them in your code. I considered deeply freezing all objects with `Object.freeze()`, but it is really slow. Then I considered doing this only in development (like [*seamless-immutable*](https://github.com/rtfeldman/seamless-immutable#performance)), but then modifying frozen objects will silently fail in development (unless you `use strict` in your code), and –worse still– succeed in production. Not good. In conclusion, **be careful** (or send me a suggestion to work around this!).

...Oh, I almost forgot! It's **tiny: just ~200 LOC and ~1.5 kB minified & compressed!**

## Benchmarks

I prepared an initial benchmarking tool comparing read/write speeds in four cases:

* In-place editing (mutable)
* [ImmutableJS](http://facebook.github.io/immutable-js/)
* [timm](https://github.com/guigrpa/timm)
* [seamless-immutable](https://github.com/rtfeldman/seamless-immutable)
* [immutability-helper](https://github.com/kolodny/immutability-helper)

All five solutions are first verified for consistency (the mutable solution obviously does not pass all tests) and then benchmarked. Benchmarks cover reading and writing object attributes at different nesting levels (root level, 2 levels and 5 levels deep), merging two small objects, and replacing an object in a 1000-long array.

Feel free to run them yourself (download the repo and then `npm install && npm run benchmarks`). These are my results on OS X for 200k iterations (Node v8.6.0):

![Benchmarks](https://github.com/guigrpa/timm/blob/master/docs/bechmarks-osx-20180822-node9.11.png?raw=true)

Some conclusions from these benchmarks:

* Reads are on par with native objects/arrays, *seamless-immutable* and *immutability-helper*, and faster than *ImmutableJS* (the deeper, the faster, even though *ImmutableJS* has improved read performance substantially in recent versions). In fact, you cannot go faster than native objects for reading!

* Writes are much slower than in-place edits, as expected, but are much faster than *seamless-immutable* (even in production mode) and *immutability-helper*, both for objects and arrays. Compared to *ImmutableJS*, object writes and merges are faster (the deeper, the faster), whereas array writes are way slower (not as slow as *seamless-immutable* and *immutability-helper*, though). For *timm* and *seamless-immutable*, write times degrade linearly with array length (and probably object size), but much more slowly for *ImmutableJS* (logarithmically?). This is where *ImmutableJS* really shines.

* Hence, what I recommend (from top to bottom):

    - If you don't need immutability, well... just **mutate in peace!** I mean, *in place*
    - If you need a complete, well-tested, rock-solid library and don't mind using a non-native API for reads: use **ImmutableJS**
    - If you value using plain arrays/objects above other considerations, use **timm**
    - If your typical use cases involve much more reading than writing, use **timm** as well
    - If you do a lot of writes, updating items in long arrays or attributes in fat objects, use **ImmutableJS**


## Usage

```js
import { merge, set as timmSet } from 'timm';
const obj = merge({ a: 2 }, { b: 3 });
const obj2 = timmSet({ foo: 1}, 'bar', 2);
```

### Arrays

#### addLast()
Returns a new array with an appended item or items.

Usage: `addLast(array, val)`

```js
arr = ['a', 'b']
arr2 = addLast(arr, 'c')
// ['a', 'b', 'c']
arr2 === arr
// false
arr3 = addLast(arr, ['c', 'd'])
// ['a', 'b', 'c', 'd']
```

#### addFirst()
Returns a new array with a prepended item or items.

Usage: `addFirst(array, val)`

```js
arr = ['a', 'b']
arr2 = addFirst(arr, 'c')
// ['c', 'a', 'b']
arr2 === arr
// false
arr3 = addFirst(arr, ['c', 'd'])
// ['c', 'd', 'a', 'b']
```

#### removeLast()
Returns a new array removing the last item.

Usage: `removeLast(array)`

```js
arr = ['a', 'b']
arr2 = removeLast(arr)
// ['a']
arr2 === arr
// false

// The same array is returned if there are no changes:
arr3 = []
removeLast(arr3) === arr3
// true
```

#### removeFirst()
Returns a new array removing the first item.

Usage: `removeFirst(array)`

```js
arr = ['a', 'b']
arr2 = removeFirst(arr)
// ['b']
arr2 === arr
// false

// The same array is returned if there are no changes:
arr3 = []
removeFirst(arr3) === arr3
// true
```

#### insert()
Returns a new array obtained by inserting an item or items
at a specified index.

Usage: `insert(array, idx, val)`

```js
arr = ['a', 'b', 'c']
arr2 = insert(arr, 1, 'd')
// ['a', 'd', 'b', 'c']
arr2 === arr
// false
insert(arr, 1, ['d', 'e'])
// ['a', 'd', 'e', 'b', 'c']
```

#### removeAt()
Returns a new array obtained by removing an item at
a specified index.

Usage: `removeAt(array, idx)`

```js
arr = ['a', 'b', 'c']
arr2 = removeAt(arr, 1)
// ['a', 'c']
arr2 === arr
// false

// The same array is returned if there are no changes:
removeAt(arr, 4) === arr
// true
```

#### replaceAt()
Returns a new array obtained by replacing an item at
a specified index. If the provided item is the same as
(*referentially equal to*) the previous item at that position,
the original array is returned.

Usage: `replaceAt(array, idx, newItem)`

```js
arr = ['a', 'b', 'c']
arr2 = replaceAt(arr, 1, 'd')
// ['a', 'd', 'c']
arr2 === arr
// false

// The same object is returned if there are no changes:
replaceAt(arr, 1, 'b') === arr
// true
```

### Collections (objects and arrays)

#### getIn()
Returns a value from an object at a given path. Works with
nested arrays and objects. If the path does not exist, it returns
`undefined`.

Usage: `getIn(obj, path)`

```js
obj = { a: 1, b: 2, d: { d1: 3, d2: 4 }, e: ['a', 'b', 'c'] }
getIn(obj, ['d', 'd1'])
// 3
getIn(obj, ['e', 1])
// 'b'
```

#### set()
Returns a new object with a modified attribute.
If the provided value is the same as (*referentially equal to*)
the previous value, the original object is returned.

Usage: `set(obj, key, val)`

```js
obj = { a: 1, b: 2, c: 3 }
obj2 = set(obj, 'b', 5)
// { a: 1, b: 5, c: 3 }
obj2 === obj
// false

// The same object is returned if there are no changes:
set(obj, 'b', 2) === obj
// true
```

#### setIn()
Returns a new object with a modified **nested** attribute.

Notes:

* If the provided value is the same as (*referentially equal to*)
the previous value, the original object is returned.
* If the path does not exist, it will be created before setting
the new value.

Usage: `setIn(obj, path, val)`

```js
obj = { a: 1, b: 2, d: { d1: 3, d2: 4 }, e: { e1: 'foo', e2: 'bar' } }
obj2 = setIn(obj, ['d', 'd1'], 4)
// { a: 1, b: 2, d: { d1: 4, d2: 4 }, e: { e1: 'foo', e2: 'bar' } }
obj2 === obj
// false
obj2.d === obj.d
// false
obj2.e === obj.e
// true

// The same object is returned if there are no changes:
obj3 = setIn(obj, ['d', 'd1'], 3)
// { a: 1, b: 2, d: { d1: 3, d2: 4 }, e: { e1: 'foo', e2: 'bar' } }
obj3 === obj
// true
obj3.d === obj.d
// true
obj3.e === obj.e
// true

// ... unknown paths create intermediate keys. Numeric segments are treated as array indices:
setIn({ a: 3 }, ['unknown', 0, 'path'], 4)
// { a: 3, unknown: [{ path: 4 }] }
```

#### update()
Returns a new object with a modified attribute,
calculated via a user-provided callback based on the current value.
If the calculated value is the same as (*referentially equal to*)
the previous value, the original object is returned.

Usage: `update(obj, key, fnUpdate)`

```js
obj = { a: 1, b: 2, c: 3 }
obj2 = update(obj, 'b', (val) => val + 1)
// { a: 1, b: 3, c: 3 }
obj2 === obj
// false

// The same object is returned if there are no changes:
update(obj, 'b', (val) => val) === obj
// true
```

#### updateIn()
Returns a new object with a modified **nested** attribute,
calculated via a user-provided callback based on the current value.
If the calculated value is the same as (*referentially equal to*)
the previous value, the original object is returned.

Usage: `updateIn<T: ArrayOrObject>(obj: T, path: Array<Key>,
fnUpdate: (prevValue: any) => any): T`

```js
obj = { a: 1, d: { d1: 3, d2: 4 } }
obj2 = updateIn(obj, ['d', 'd1'], (val) => val + 1)
// { a: 1, d: { d1: 4, d2: 4 } }
obj2 === obj
// false

// The same object is returned if there are no changes:
obj3 = updateIn(obj, ['d', 'd1'], (val) => val)
// { a: 1, d: { d1: 3, d2: 4 } }
obj3 === obj
// true
```

#### merge()
Returns a new object built as follows: the overlapping keys from the
second one overwrite the corresponding entries from the first one.
Similar to `Object.assign()`, but immutable.

Usage:

* `merge(obj1, obj2)`
* `merge(obj1, ...objects)`

The unmodified `obj1` is returned if `obj2` does not *provide something
new to* `obj1`, i.e. if either of the following
conditions are true:

* `obj2` is `null` or `undefined`
* `obj2` is an object, but it is empty
* All attributes of `obj2` are `undefined`
* All attributes of `obj2` are referentially equal to the
  corresponding attributes of `obj1`

Note that `undefined` attributes in `obj2` do not modify the
corresponding attributes in `obj1`.

```js
obj1 = { a: 1, b: 2, c: 3 }
obj2 = { c: 4, d: 5 }
obj3 = merge(obj1, obj2)
// { a: 1, b: 2, c: 4, d: 5 }
obj3 === obj1
// false

// The same object is returned if there are no changes:
merge(obj1, { c: 3 }) === obj1
// true
```

#### mergeDeep()
Returns a new object built as follows: the overlapping keys from the
second one overwrite the corresponding entries from the first one.
If both the first and second entries are objects they are merged recursively.
Similar to `Object.assign()`, but immutable, and deeply merging.

Usage:

* `mergeDeep(obj1, obj2)`
* `mergeDeep(obj1, ...objects)`

The unmodified `obj1` is returned if `obj2` does not *provide something
new to* `obj1`, i.e. if either of the following
conditions are true:

* `obj2` is `null` or `undefined`
* `obj2` is an object, but it is empty
* All attributes of `obj2` are `undefined`
* All attributes of `obj2` are referentially equal to the
  corresponding attributes of `obj1`

Note that `undefined` attributes in `obj2` do not modify the
corresponding attributes in `obj1`.

```js
obj1 = { a: 1, b: 2, c: { a: 1 } }
obj2 = { b: 3, c: { b: 2 } }
obj3 = mergeDeep(obj1, obj2)
// { a: 1, b: 3, c: { a: 1, b: 2 }  }
obj3 === obj1
// false

// The same object is returned if there are no changes:
mergeDeep(obj1, { c: { a: 1 } }) === obj1
// true
```

#### mergeIn()
Similar to `merge()`, but merging the value at a given nested path.

Usage examples:

* `mergeIn(obj1, path, obj2)`
* `mergeIn(obj1, path, ...objects)`

```js
obj1 = { a: 1, d: { b: { d1: 3, d2: 4 } } }
obj2 = { d3: 5 }
obj3 = mergeIn(obj1, ['d', 'b'], obj2)
// { a: 1, d: { b: { d1: 3, d2: 4, d3: 5 } } }
obj3 === obj1
// false

// The same object is returned if there are no changes:
mergeIn(obj1, ['d', 'b'], { d2: 4 }) === obj1
// true
```

#### omit()
Returns an object excluding one or several attributes.

Usage: `omit(obj, attrs)`

```js
obj = { a: 1, b: 2, c: 3, d: 4 }
omit(obj, 'a')
// { b: 2, c: 3, d: 4 }
omit(obj, ['b', 'c'])
// { a: 1, d: 4 }

// The same object is returned if there are no changes:
omit(obj, 'z') === obj1
// true
```

#### addDefaults()
Returns a new object built as follows: `undefined` keys in the first one
are filled in with the corresponding values from the second one
(even if they are `null`).

Usage:

* `addDefaults(obj, defaults)`
* `addDefaults(obj, ...defaultObjects)`

```js
obj1 = { a: 1, b: 2, c: 3 }
obj2 = { c: 4, d: 5, e: null }
obj3 = addDefaults(obj1, obj2)
// { a: 1, b: 2, c: 3, d: 5, e: null }
obj3 === obj1
// false

// The same object is returned if there are no changes:
addDefaults(obj1, { c: 4 }) === obj1
// true
```


## MIT license

Copyright (c) [Guillermo Grau Panea](https://github.com/guigrpa) 2016-present

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
* Improve TypeScript signatures (fix overload order + add missing catch-alls).

## 1.7.0 (2020-8-27)

* Rewritten in TypeScript.

## 1.6.2 (2019-8-11)

* Update package metadata.
* Update deps to fix security audit.

## 1.6.1 (2018-8-8)

* Update deps to fix security audit.

## 1.6.0 (2018-8-8)

* Replace functions when merging (`merge()` and `mergeDeep()`) [#25, @hipstersmoothie]

## 1.5.0 (2018-7-23)

* Add TypeScript definitions [#23, @grrowl]

## 1.4.2 (2018-3-29)

* Support Symbol properties in objects.

## 1.4.1 (2018-1-2)

* Add `mergeDeep()` benchmarks [#16, @dentrado]
* doSetIn: ensure nested value is object [#17, @petermikitsh]

## 1.4.0 (2017-10-25)

* Add **mergeDeep()** [#14, @dentrado]
* Update benchmarks to reflect current versions of *ImmutableJS* and *seamless-immutable*

## 1.3.1 (2017-10-14)

* Add **update()** to Flow API [#13, @dentrado]

## 1.3.0 (October 3, 2017)

* Add **update()** [#12, @dentrado]

## 1.2.5 (April 2, 2017)

* Improve `replaceAt` performance [#9, @gamtiq]

## 1.2.4 (March 4, 2017)

* Bugfix: Fix `setIn` object creation for arrays in some edge cases [#8, @adierkens]

## 1.2.3 (December 8, 2016)

* Bugfix: Allow `set`/`setIn` to automatically create arrays, handling correctly negative and decimal numbers [#4, @YellowKirby]

## 1.2.2 (December 7, 2016)

* Bugfix: Allow `set`/`setIn` to automatically create arrays [#3, @YellowKirby]

## 1.2.1 (November 15, 2016)

* Update external Flow type definitions to support Flex 0.35.

## 1.2.0 (October 24, 2016)

* Add **removeFirst(), removeLast()**

## 1.1.3 (October 19, 2016)

* Make user-facing Flow type definitions inherit from internal ones (to avoid divergence).

## 1.1.2 (October 19, 2016)

* Change how Flow type definitions are exposed so they can be used with zero config.

## 1.1.1 (October 18, 2016)

* Fix Flow type definitions for object-related function outputs, making them more flexible.

## 1.1.0 (October 10, 2016)

* Expose **Flow** type definitions.

## 1.0.0 (June 13, 2016)

* **Use semantic versioning**
* [minor] Bump development dependencies.

## 0.6.1 (Apr. 22, 2016)

* Add `omit()`

## 0.6.0 (Mar. 24, 2016)

* Renamed build directory to `lib`.

## 0.5.3 (Mar. 23, 2016)

* Removed Babel config (.babelrc) from the published package (npm). This simplifies consumption of the ES6 (or ES6 with Flow) versions.

## 0.5.1, 0.5.2 (Mar. 23, 2016)

* Distribute 4 versions of Timm:
    - ES5 (default): `timm.js` (just `require('timm')`)
    - ES5 minimised: `timm.min.js`
    - ES6: `timm_es6.js`
    - ES6 with Flow type annotations: `timm_es6_flow.js`

## 0.5.0 (Mar. 23, 2016)

* Important internal changes (migrated to AVA, ES6, Flow), but usage remains unchanged

## 0.4.3 (Mar. 08, 2016)

* Bugfix: internal `_merge()` incorrectly merged `{a: 3} + {a: undefined} => {a: undefined}`

## 0.4.2 (Feb. 28, 2016)

* Update documentation for the new functions added in 0.4.1.

## 0.4.1 (Feb. 28, 2016)

* Add `getIn()`, `updateIn()`, `mergeIn()`
* `setIn()` now creates nested objects for unknown paths
* Bugfix: internal `_clone()` converted arrays to objects

## 0.4.0 (Feb. 25, 2016)

* Accept arrays in array operations `addLast()` and `addFirst()`.
* Add new array operation `insert()`.


tools/benchmarks.js
process.env.NODE_ENV = 'production';
const _ = require('lodash');
const chalk = require('chalk');
// const Seamless = require('seamless-immutable')
const Seamless = require('seamless-immutable/seamless-immutable.production.min');
const Immutable = require('immutable');
const update = require('immutability-helper');
const timm = require('../lib/timm');

const INITIAL_OBJECT = {
  toggle: false,
  b: 3,
  str: 'foo',
  d: {
    d1: 6,
    d2: 'foo',
    toggle: false,
    d9: { b: { b: { b: 1 } } },
  },
  e: {
    e1: 18,
    e2: 'foo',
  },
};
const DEEP_PATH = ['d', 'd9', 'b', 'b', 'b'];
const ARRAY_LENGTH = 1000;
const INITIAL_ARRAY = new Array(ARRAY_LENGTH);
for (let n = 0; n < ARRAY_LENGTH; n++) {
  INITIAL_ARRAY[n] = { a: 1, b: 2 };
}
const N = 2e5;

const _getIn = (obj, path) => {
  let out = obj;
  path.forEach(key => {
    out = out[key];
  });
  return out;
};

const _mergeDeepInPlace = (obj1, obj2) => {
  Object.keys(obj2).forEach(key => {
    const val = obj2[key];
    if (val == null) obj1[key] = val;
    else if (typeof val === 'object' && val.length != null)
      obj1[key] = _mergeDeepInPlace([], val);
    else if (typeof val === 'object') obj1[key] = _mergeDeepInPlace({}, val);
    else obj1[key] = val;
  });
  return obj1;
};

const _solMutable = {
  init: () => _.cloneDeep(INITIAL_OBJECT),
  get: (obj, key) => obj[key],
  set: (obj, key, val) => {
    obj[key] = val;
    return obj;
  },
  getDeep: (obj, key1, key2) => obj[key1][key2],
  setDeep: (obj, key1, key2, val) => {
    obj[key1][key2] = val;
    return obj;
  },
  getIn: _getIn,
  setIn: (obj, path, val) => {
    let ptr = obj;
    for (let idx = 0; idx < path.length - 1; idx++) {
      ptr = ptr[path[idx]];
    }
    ptr[path[path.length - 1]] = val;
    return obj;
  },
  merge: (obj1, obj2) => {
    Object.keys(obj2).forEach(key => {
      obj1[key] = obj2[key];
    });
    return obj1;
  },
  mergeDeep: (obj1, obj2) => _mergeDeepInPlace(obj1, obj2),
  initArr: () => _.cloneDeep(INITIAL_ARRAY),
  getAt: (arr, idx) => arr[idx],
  setAt: (arr, idx, val) => {
    arr[idx] = val;
    return arr;
  },
};

_solImmutableTimm = {
  init: () => _.cloneDeep(INITIAL_OBJECT),
  get: (obj, key) => obj[key],
  set: (obj, key, val) => timm.set(obj, key, val),
  getDeep: (obj, key1, key2) => obj[key1][key2],
  setDeep: (obj, key1, key2, val) =>
    timm.set(obj, key1, timm.set(obj[key1], key2, val)),
  getIn: _getIn,
  setIn: (obj, path, val) => timm.setIn(obj, path, val),
  merge: (obj1, obj2) => timm.merge(obj1, obj2),
  mergeDeep: (obj1, obj2) => timm.mergeDeep(obj1, obj2),
  initArr: () => _.cloneDeep(INITIAL_ARRAY),
  getAt: (arr, idx) => arr[idx],
  setAt: (arr, idx, val) => timm.replaceAt(arr, idx, val),
};

_solImmutableJs = {
  init: () => Immutable.fromJS(INITIAL_OBJECT), // deep
  get: (obj, key) => obj.get(key),
  set: (obj, key, val) => obj.set(key, val),
  getDeep: (obj, key1, key2) => obj.getIn([key1, key2]),
  setDeep: (obj, key1, key2, val) => obj.setIn([key1, key2], val),
  getIn: (obj, path) => obj.getIn(path),
  setIn: (obj, path, val) => obj.setIn(path, val),
  merge: (obj1, obj2) => obj1.merge(obj2),
  mergeDeep: (obj1, obj2) => obj1.mergeDeep(obj2),
  initArr: () => Immutable.List(INITIAL_ARRAY), // shallow
  getAt: (arr, idx) => arr.get(idx),
  setAt: (arr, idx, val) => arr.set(idx, val),
};

_solImmutableSeamless = {
  init: () => Seamless(INITIAL_OBJECT),
  get: (obj, key) => obj[key],
  set: (obj, key, val) => obj.set(key, val),
  getDeep: (obj, key1, key2) => obj[key1][key2],
  setDeep: (obj, key1, key2, val) => obj.setIn([key1, key2], val),
  getIn: _getIn,
  setIn: (obj, path, val) => obj.setIn(path, val),
  merge: (obj1, obj2) => obj1.merge(obj2),
  mergeDeep: (obj1, obj2) => obj1.merge(obj2, { deep: true }),
  initArr: () => Seamless(INITIAL_ARRAY),
  getAt: (arr, idx) => arr[idx],
  setAt: (arr, idx, val) => arr.set(idx, val),
};

_solImmutableUpdate = {
  init: () => _.cloneDeep(INITIAL_OBJECT),
  get: (obj, key) => obj[key],
  set: (obj, key, val) => {
    const config = {};
    config[key] = { $set: val };
    return update(obj, config);
  },
  getDeep: (obj, key1, key2) => obj[key1][key2],
  setDeep: (obj, key1, key2, val) => {
    const config = {};
    config[key1] = {};
    config[key1][key2] = { $set: val };
    return update(obj, config);
  },
  getIn: _getIn,
  setIn: (obj, path, val) => {
    const config = {};
    const len = path.length;
    let child = config;
    for (let n = 0; n < len; n++) {
      child = child[path[n]] = n === len - 1 ? { $set: val } : {};
    }
    return update(obj, config);
  },
  merge: (obj1, obj2) => update(obj1, { $merge: obj2 }),
  mergeDeep: (obj1, obj2) => update(obj1, _nestedConfig(obj2)),
  initArr: () => _.cloneDeep(INITIAL_ARRAY),
  getAt: (arr, idx) => arr[idx],
  setAt: (arr, idx, val) => {
    const config = {};
    config[idx] = { $set: val };
    return update(arr, config);
  },
};

const _nestedConfig = _.memoize(obj => {
  return Object.keys(obj).reduce((result, key) => {
    const val = obj[key];
    if (typeof val === 'object' && val != null) {
      result[key] = _nestedConfig(val);
    } else {
      result[key] = { $set: val };
    }
    return result;
  }, {});
});

const _toggle = (solution, obj) =>
  solution.set(obj, 'toggle', !solution.get(obj, 'toggle'));

const _addResult = (results, condition) => {
  results.push(condition ? chalk.green.bold('P') : chalk.red('F'));
};

const _verify = solution => {
  const results = [];
  const {
    init,
    get,
    set,
    setDeep,
    getIn,
    setIn,
    merge,
    mergeDeep,
    initArr,
    getAt,
    setAt,
  } = solution;
  let obj;
  let obj2;
  let arr;
  let arr2;

  // Initial conditions
  obj = init();
  _addResult(results, get(obj, 'toggle') === false);
  results.push('-');

  // Changes to root attributes create a new object
  // (but keep the nested object untouched)
  obj2 = set(obj, 'toggle', true);
  _addResult(results, get(obj, 'toggle') === false);
  _addResult(results, get(obj2, 'toggle') === true);
  _addResult(results, obj2 !== obj);
  _addResult(results, get(obj2, 'd') === get(obj, 'd'));
  results.push('-');
  obj2 = set(obj, 'str', 'foo');
  _addResult(results, obj2 === obj);
  _addResult(results, get(obj2, 'd') === get(obj, 'd'));
  results.push('-');

  // Same for deep attributes
  obj2 = setDeep(obj, 'd', 'd1', 3);
  _addResult(results, solution.getDeep(obj, 'd', 'd1') === 6);
  _addResult(results, solution.getDeep(obj2, 'd', 'd1') === 3);
  _addResult(results, obj2 !== obj);
  _addResult(results, get(obj2, 'd') !== get(obj, 'd'));
  _addResult(results, get(obj2, 'e') === get(obj, 'e'));
  results.push('-');

  // If we change an attribute to exactly the same value,
  // no new object is created
  obj2 = set(obj, 'b', get(obj, 'b'));
  _addResult(results, obj2 === obj);
  _addResult(results, get(obj2, 'd') === get(obj, 'd'));
  results.push('-');
  obj2 = set(obj, 'str', 'bar');
  _addResult(results, obj2 !== obj);
  _addResult(results, get(obj2, 'd') === get(obj, 'd'));

  // Same for deep attributes
  obj = init();
  obj2 = setDeep(obj, 'd', 'd1', 6);
  _addResult(results, solution.getDeep(obj, 'd', 'd1') === 6);
  _addResult(results, solution.getDeep(obj2, 'd', 'd1') === 6);
  _addResult(results, obj2 === obj);
  _addResult(results, get(obj2, 'd') === get(obj, 'd'));
  results.push('-');

  // Deep writes
  obj2 = setIn(obj, DEEP_PATH, 3);
  _addResult(results, obj2 !== obj);
  _addResult(results, get(obj2, 'd') !== get(obj, 'd'));
  _addResult(results, get(obj2, 'e') === get(obj, 'e'));
  _addResult(results, getIn(obj, DEEP_PATH) === 1);
  _addResult(results, getIn(obj2, DEEP_PATH) === 3);
  results.push('-');

  // Merge
  obj2 = merge(obj, { c: 5, f: null });
  _addResult(results, obj2 !== obj);
  _addResult(results, get(obj2, 'd') === get(obj, 'd'));
  _addResult(results, get(obj2, 'c') === 5);
  _addResult(results, get(obj2, 'f') === null);
  results.push('-');

  // Deep merge
  obj2 = mergeDeep(obj, { c: 5, f: null, d: { d9: { b: { a: 1 } } } });
  const changedPath = ['d', 'd9', 'b', 'a'];
  const unchangedPath = ['d', 'd9', 'b', 'b'];
  _addResult(results, obj2 !== obj);
  _addResult(results, get(obj2, 'd') !== get(obj, 'd'));
  _addResult(results, getIn(obj2, changedPath) === 1);
  _addResult(results, getIn(obj2, unchangedPath) === getIn(obj, unchangedPath));
  _addResult(results, get(obj2, 'c') === 5);
  _addResult(results, get(obj2, 'f') === null);
  results.push('-');

  // Array writes
  arr = initArr();
  arr2 = setAt(arr, 1, { b: 3 });
  _addResult(results, arr2 !== arr);
  _addResult(results, getAt(arr, 1).b === 2);
  _addResult(results, getAt(arr2, 1).b === 3);
  arr2 = setAt(arr, 1, getAt(arr, 1));
  _addResult(results, arr2 === arr);

  console.log(`  Verification: ${results.join('')}`);
};

const _test = (desc, cb) => {
  const tic = new Date().getTime();
  cb();
  const tac = new Date().getTime();
  console.log(`  ${desc}: ` + chalk.bold(`${tac - tic} ms`));
};

const _allTests = (desc, solution) => {
  console.log(chalk.bold(desc));
  _verify(solution);
  obj = solution.init();
  _test(`Object: read (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const val = solution.get(obj, 'toggle');
    }
  });
  obj = solution.init();
  _test(`Object: write (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const obj2 = solution.set(obj, 'b', n);
    }
  });
  obj = solution.init();
  _test(`Object: deep read (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const val = solution.getDeep(obj, 'd', 'd1');
    }
  });
  obj = solution.init();
  _test(`Object: deep write (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const obj2 = solution.setDeep(obj, 'd', 'd1', n);
    }
  });
  obj = solution.init();
  _test(`Object: very deep read (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const val = solution.getIn(obj, DEEP_PATH);
    }
  });
  obj = solution.init();
  _test(`Object: very deep write (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const obj2 = solution.setIn(obj, DEEP_PATH, n);
    }
  });
  obj = solution.init();
  MERGE_OBJ = { c: 5, f: null };
  _test(`Object: merge (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const obj2 = solution.merge(obj, MERGE_OBJ);
    }
  });
  obj = solution.init();
  MERGE_DEEP_OBJ = { c: 5, f: null, d: { d9: { b: { a: 1 } } } };
  _test(`Object: deep merge (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const obj2 = solution.mergeDeep(obj, MERGE_DEEP_OBJ);
    }
  });
  arr = solution.initArr();
  _test(`Array: read (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const val = solution.getAt(arr, 1);
    }
  });
  arr = solution.initArr();
  _test(`Array: write (x${N})`, () => {
    for (let n = 0; n < N; n++) {
      const arr2 = solution.setAt(arr, 1, n);
    }
  });
};

_allTests('Mutable', _solMutable);
_allTests('Immutable (ImmutableJS)', _solImmutableJs);
_allTests('Immutable (timm)', _solImmutableTimm);
_allTests('Immutable (seamless-immutable)', _solImmutableSeamless);
_allTests('Immutable (immutability-helper)', _solImmutableUpdate);
